src/main/java/net/rptools/maptool/client/ui/htmlframe/HTMLContent.java

```diff
@@ -0,0 +1,418 @@
+  public boolean isUrl() {
+    return url != null;
+  }
+
+  /**
+   * Checks if the HTML content is a string.
+   *
+   * @return true if the HTML content is a string, false otherwise
+   */
+  public boolean isHTMLString() {
+    return htmlString != null;
+  }
+
+  /**
+   * Returns the HTML content as a string. If the content is a URL it will return null.
+   *
+   * @return the HTML content as a string
+   */
+  public String getHtmlString() {
+    return htmlString;
+  }
+
+  /**
+   * Returns the URL of the HTML content. If the content is a string it will return null.
+   *
+   * @return the URL of the HTML content
+   */
+  public URL getUrl() {
+    return url;
+  }
+
+  /**
+   * Returns the HTML content as a data URL in base64 format.
+   *
+   * @return the HTML content as a data URL in base64 format.
+   */
+  public String getHtmlStringAsDataUrl() {
+    if (isHTMLString()) {
+      String encodedHtml =
+          Base64.getEncoder().encodeToString(htmlString.getBytes(StandardCharsets.UTF_8));
+      return "data:text/html;base64," + encodedHtml;
+    } else {
+      throw new IllegalStateException("HTMLContent is not a string");
+    }
+  }
+
+  /**
+   * Injects the base URL tag into the HTML content if it is a string.
+   *
+   * @return the HTMLContent object with the base URL injected.
+   */
+  public HTMLContent injectURLBase(@Nonnull URL baseUrl) {
+    if (baseUrlInjected) {
+      return this; // already injected so return the same object
+    }
+    if (isHTMLString()) {
+      return new HTMLContent(injectURLBase(htmlString, baseUrl), url, javaBridgeInjected, true);
+    } else {
+      throw new IllegalStateException("HTMLContent is not a string");
+    }
+  }
+
+  /**
+   * Parses the HTML in the string and sets the base to the correct location.
+   *
+   * @param htmlString the HTML to parse.
+   * @param url the origin URL to set the base relative to
+   * @return a string containing the HTML with the base URL injected.
+   */
+  public String injectURLBase(@Nonnull String htmlString, @Nonnull URL url) {
+    var document = Jsoup.parse(htmlString);
+    var head = document.select("head").first();
+    if (head != null) {
+      addBase(document, url);
+    }
+    return document.html();
+  }
+
+  /**
+   * Injects the Java Bridge and the base URL into the HTML content if it is a string. If the
+   * content is a URL it will return the same object without any changes. If the content has already
+   * been injected it will return the same object.
+   *
+   * <p>If url is null, the base URL will not be injected.
+   *
+   * @param url the URL of the HTML content
+   * @return the HTMLContent object with the Java Bridge ad base URL injected.
+   */
+  public HTMLContent injectJavaBridgeAndBaseUrl(@Nullable URL url) {
+    if (javaBridgeInjected && baseUrlInjected) {
+      return this; // both already injected so return the same object
+    }
+    var newHtml = htmlString;
+    var document = Jsoup.parse(newHtml);
```

Comment:
Chance for NPE here as `htmlString` could be `null` at this point.

Reply:
Now throws an IllegalStateException